File: Hallway.java
 /*
 * @author Renzo Svartz
 */

package src;

import src.Main.ORIENTATION;

/**
 * Class implementing a vertex
 */
public class Hallway
{
	//hallway_name, vertex_a, cost_a,  vertex_b, cost_b
	
	private String hallway_coor;
	
    private Custom_Vertex vertex_a;
    private int cost_a;
    private ORIENTATION to_a;
    private ORIENTATION from_a;

    private Custom_Vertex vertex_b;
    private int cost_b;
    private ORIENTATION to_b;
    private ORIENTATION from_b;

	private boolean visited;
	
	//get_name, visit, unvisit, is_visited, get set has predecessor, get set Cost
	
	/**
	 * Parameterized Constructor
	 * @param vertex_name vertex name
	 */
	public Hallway(String hallway_coor, Custom_Vertex vertex_a, int cost_a, ORIENTATION to_a, ORIENTATION from_a)
	{
		this.hallway_coor = hallway_coor;
        this.vertex_a = vertex_a;
		this.cost_a = cost_a;
        this.to_a = to_a;
        this.from_a = from_a;

        this.vertex_b = null;
        this.cost_b = Integer.MAX_VALUE;
        this.to_b = null;
        this.from_b = null;

        this.visited = false;
	}

    public Custom_Vertex get_a()
    {
        return vertex_a;
    }

	/** Gets the recorded cost_a of the path to this vertex.
	@return The cost_a of the path. */
	public int get_cost_a()
	{
		return cost_a;
	}

    public ORIENTATION get_to_a() 
	{
		return to_a;
	}

    public ORIENTATION get_from_a() 
	{
		return from_a;
	}

    public void add_vertex_b(Custom_Vertex vertex_b, int cost_b, ORIENTATION to_b, ORIENTATION from_b) 
	{
		this.vertex_b = vertex_b;
        this.cost_b = cost_b;
        this.to_b = to_b;
        this.from_b = from_b;
	}

    public Custom_Vertex get_b()
    {
        return vertex_b;
    }

	/** Gets the recorded cost_a of the path to this vertex.
	@return The cost_a of the path. */
	public int get_cost_b()
	{
		return cost_b;
	}

    public ORIENTATION get_to_b() 
	{
		return to_b;
	}

    public ORIENTATION get_from_b() 
	{
		return from_b;
	}
	
    /** Marks this vertex as visited. */
	public void visit()
	{
		visited = true;
	}
	
	/** Sees whether this vertex is marked as visited.
	@return True if the vertex is visited. */
	public boolean is_visited()
	{
		return visited;
	}

	public String toString()
	{
		return hallway_coor;
	}
}

File: Hallway.c
#include "Hallway.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

Hallway* createHallway(char* hallway_coor, Custom_Vertex* vertex_a, int cost_a, ORIENTATION to_a, ORIENTATION from_a) {
    Hallway* newHallway = (Hallway*)malloc(sizeof(Hallway));
    if (newHallway == NULL) {
        fprintf(stderr, "Error: Memory allocation failed\n");
        exit(EXIT_FAILURE);
    }
    newHallway->hallway_coor = strdup(hallway_coor);
    newHallway->vertex_a = vertex_a;
    newHallway->cost_a = cost_a;
    newHallway->to_a = to_a;
    newHallway->from_a = from_a;
    newHallway->vertex_b = NULL;
    newHallway->cost_b = INT_MAX;
    newHallway->to_b = NULL;
    newHallway->from_b = NULL;
    newHallway->visited = false;
    return newHallway;
}

Custom_Vertex* get_a(Hallway* hallway) {
    return hallway->vertex_a;
}

int get_cost_a(Hallway* hallway) {
    return hallway->cost_a;
}

ORIENTATION get_to_a(Hallway* hallway) {
    return hallway->to_a;
}

ORIENTATION get_from_a(Hallway* hallway) {
    return hallway->from_a;
}

void add_vertex_b(Hallway* hallway, Custom_Vertex* vertex_b, int cost_b, ORIENTATION to_b, ORIENTATION from_b) {
    hallway->vertex_b = vertex_b;
    hallway->cost_b = cost_b;
    hallway->to_b = to_b;
    hallway->from_b = from_b;
}

Custom_Vertex* get_b(Hallway* hallway) {
    return hallway->vertex_b;
}

int get_cost_b(Hallway* hallway) {
    return hallway->cost_b;
}

ORIENTATION get_to_b(Hallway* hallway) {
    return hallway->to_b;
}

ORIENTATION get_from_b(Hallway* hallway) {
    return hallway->from_b;
}

void visit(Hallway* hallway) {
    hallway->visited = true;
}

bool is_visited(Hallway* hallway) {
    return hallway->visited;
}

char* toString(Hallway* hallway) {
    return hallway->hallway_coor;
}

void destroyHallway(Hallway* hallway) {
    free(hallway->hallway_coor);
    free(hallway);
}
